### DISCUSSION TITLE
Extra line break when assigning htmlText to a TextField when TextFieldType is INPUT.

### DISCUSSION THREAD

**Discussion by manishpatel0404** [airsdk/Adobe-Runtime-Support#2065] [Q&A]
Has anyone ever found a solution to this [StackOverFlow link?](https://stackoverflow.com/a/21316962)

**Comment by ajwfrost** [comment]
I'm not sure that is actually a bug ... if you have a text field with contents `<p>Hello</p>` and then want to add any text to the end of the string, it will be on the next line. So the newline character is valid: certainly, when you are editing the text, you would need to have the extra line available so that you can type there. So if anything, the problem is that this is stripped off when the text field is set to read-only/dynamic mode.

The solution if you don't want to have that extra newline would presumably be to switch it to dynamic mode when the focus is lost, and then back to input mode if it's clicked?
```
			var tf : TextField = new TextField();
			tf.type = TextFieldType.INPUT;
			tf.autoSize = TextFieldAutoSize.LEFT;
			tf.multiline = true;
			tf.border = true;
			tf.borderColor = 0;
			addChild(tf);
			tf.htmlText = "<p align='right'>abcde</p>";
			tf.addEventListener(FocusEvent.FOCUS_OUT, function(e:Event):void {
				tf.type = TextFieldType.DYNAMIC;
			});
			tf.addEventListener(MouseEvent.CLICK, function(e:Event):void {
				if (tf.type == TextFieldType.DYNAMIC) tf.type = TextFieldType.INPUT;
			});
```

Alternatively we could fairly easily change this so that the extra newline (and associated resizing of the box) is only there when you have focus. So this would mean the input field is changing size depending on whether you're clicked in it or not. If you assigned that `<p>Hello</p>` to it as an Input type, it would only show as a single line height, but when you clicked on it to focus it, it would resize back to two lines... I'm a little hesitant to change the behaviour of this though as it's been there for a while and is fairly deliberate.

**Reply by manishpatel0404** [reply]
@ajwfrost, I agree that this is not a bug, but more of a design-level thought. However, when I am editing text, I would not mind hitting the Return/Enter key to create the new line. This way, at least the same `TextFormat` is retained, as the earlier paragraph. This is also analogous with most Word Processors, which end-users are habituated with.

In the current scenario, the `TextFormat` of the newline is set to the `defaultTextFormat`, which is what was flagged by our users, who then had to re-select the formatting from our custom toolbar. Alternatively, they had to hit the BackSpace to delete the new line and then hit the Return/Enter key to retain the earlier paragraph's `TextFormat`. Imagine having to do this in each of the `multiline` TextFields in a data-entry form that contains multilingual html text.

We save the data in a MySQL database and every time the data is fetched from the db for editing, each of the TextFields introduce the extra newline. I did follow the advice in same SO article I have referenced in my post, by stripping the last `</P>` tag from the html before assigning it to to the TextField. This got rid of the newline, while maintaining the `TextFormat`. Thankfully, when saving the data back to the database, the `</P>` tag gets added back. So I stuck with the workaround and my end-users see what they are used to. Although the workaround works, from a programming standpoint, I found it a bit cheeky.

Also, to create event listeners to solve this would create unnecessary bloat in the code when there are many TextFields in a form, even if they all have a common event handler.

IMHO, let users hit the Return/Enter key if they want a new line. This is in line with how most Word Processors work.

On the whole, I have re-written 6 of our Company's Lingo-based applications, including an internal UI framework that is used by these applications in AS3 over the last 2 years and have yet to come across a situation where I had to scream 'bug'. Kudos to you and your team for taking up this challenge and more power to you all.

Thanks.

**Reply by ajwfrost** [reply]
Hi - interesting, so with your comment about the extra line being set to the `defaultTextFormat`, I've just had a check and yes it seems that the field is interpreted as a paragraph with the given style, and then with the `</p>` it is closing that style and then adding a newline. But of course, this is exactly what would happen in HTML, e.g. `<p align="right">Hello</p>world`

From your description then, I gather you're using the paragraph tag in order to set the formatting, and would like this formatting to be applied without the closing paragraph tag causing a newline (and so that if someone clicks at the end of the text, they're essentially clicking prior to the `</p>` mark so that they continue with the same paragraph format...

Some interesting observations:
If I set the html text to `"<p align='right'>abcde</p>"` and then trace out the values of `text` and `htmlText` (each enclosed in square brackets), I get:
```
Text = [abcde
]
HTML = [<P ALIGN="RIGHT"><FONT FACE="Times New Roman" SIZE="12" COLOR="#000000" LETTERSPACING="0" KERNING="0">abcde</FONT></P>]
```
and if I then strip off the newline character from the `text` value, the `htmlText` ends up as
```
HTML = [<P ALIGN="LEFT"><FONT FACE="Times New Roman" SIZE="12" COLOR="#000000" LETTERSPACING="0" KERNING="0">abcde</FONT></P>]
```
If I do the same with just  `"<p>abcde</p>"` then the `htmlText` is just
```
HTML = [<P ALIGN="LEFT"><FONT FACE="Times New Roman" SIZE="12" COLOR="#000000" LETTERSPACING="0" KERNING="0">abcde</FONT></P>]
```
both before and after I strip the newline - and after stripping the newline, the text field is only a single line in height.

So I guess this is doing roughly what you're saying: perhaps stripping the newline is equivalent to stripping the final `</p>` but then when you read it back, it gets added again...

I still have a concern that this is the behaviour that is correct, both from the HTML perspective and also from a "how else could you do this" perspective (i.e. if you wanted to have the behaviour where the first line has right alignment and then the second line has left alignment, this is how it would need to be done...). I'm wondering whether what you're requiring would be better achieved with `<span>` and a style sheet, or perhaps if all you're looking at is the text alignment (and assuming you want to keep it all self-contained rather than using separate style sheets), then a simpler option may be if we added the option for alignment to be included in the `<textformat>` tag, that may work? Although of course, alignment is by nature a paragraph format, so it would be a problem if we had multiple of these instructions within a single line -> we'd probably just take the latest value (having seen how the code parses these!)...

But yes, from looking at the code and how the `htmlText` is being parsed, it's possibly best to leave off the final `</p>` of the html since when it encounters this, the text field is popping the character format to that of the previous paragraph/style, and adding a newline. Changing this behaviour would probably cause a number of regression issues for other folk...

thanks

**Reply by manishpatel0404** [reply]
> But of course, this is exactly what would happen in HTML, e.g. <p align="right">Hello</p>world

But what would happen if you type the following in an INPUT type `TextField` - **Hello** (notice the bold formatting) followed by hitting Return (newline) and then type "world" on the new line. Would the word "world" not also be bold? This does not happen with the manually inserted extra newline when applying `htmlText` to the `TextField`. The extra newline reverts to the `defaultTextFormat`. If the manually inserted extra newline is made to carry forward the `TextFormat` of the previous character, the behaviour would be in line with most Word Processors. Since I don't have access to the code, I am not sure, but I think, currently the newline does not carry forward the `TextFormat` of the previous character. I think the confusion here is because of the differing perceptions of rendering Html Text vs Rich Text. Generally, I believe, the `htmlText` that one would set in a `TextField` would be perceived as Rich Text with at least some text formatting applied to it and not plain-vanilla html. Even the example above contains formatted text, namely alignment.

> From your description then, I gather you're using the paragraph tag in order to set the formatting, and would like this formatting to be applied without the closing paragraph tag causing a newline...

Not really. This is the default behaviour of an INPUT type `TextField`, in that it uses the `TextFormat` of the last character for the next character unless it is changed through code or some UI process. So, the `<P>` tag has nothing to do with this. Let me explain our process in a bit more detail. We have some TextFields that require descriptive information that contains Rich Text formatting such as different fonts, styles and colors. To achieve this, we have set up a formatting toolbar UI that works in conjunction with these `multiline` TextFields. While typing text, the user selects the desired formatting. When the Save button is clicked, we save the html, so as to retain the `TextFormat` of the entire text. When the text is retrieved and set into a `TextField` as htmlText, the extra new line reverts to the `defaultTextFormat`. This then, we have to delete and hit Return again in the INPUT mode to continue the existing `TextFormat`. To avoid this, we stripped off the last `</P>` tag. By the way, the moment you hit Return, the stripped `</P>` tag gets 'automagically' added back, which is what saved my day. And the earlier `TextFormat` is also retained. So the users were happy.

But, I don't think there is any magic in this. You see, the text is being typed in an INPUT type `TextField`. We collect the html only when the user clicks the Save button, at which time, I believe, the html text is re-generated by the `TextField` from the `TextFormat` of the characters in the text.

> i.e. if you wanted to have the behaviour where the first line has right alignment and then the second line has left alignment, this is how it would need to be done...

The extra new line is inserted only when `tf.type = TextFieldType.INPUT`. In input mode, you would require either a formatting UI toolbar or Keyboard shortcuts (such as Ctrl-B, etc.) to do do any sort of text formatting, including text alignment.

> Changing this behaviour would probably cause a number of regression issues for other folk...

Or you could simply add a new `public` property to the `TextField`, something that represents "add a newline when setting htmlText in input mode" and keep its default to `true` to maintain backward-compatibility. And no need to strip the final `</P>` either. Best of both worlds, maybe?

In any case, stripping off the `</P>` tag has worked for us and got us the desired results.

Thanks.

**Reply by ajwfrost** [reply]
Internally the way it works is a set of styles/text formats that are parsed from the `htmlText` value, with paragraph formatting such as alignment being pushed/popped when it encounters the `<p>` and `</p>` tags. So that is why you'd see the default text format coming back after the `</p>` which is at the end of the paragraph (hence a newline). In other words, if you've set some text into the text field using `htmlText` and it's enclosed by paragraph tags, then clicking at the end of the text (or within the second line) means your insertion point is _after_ the newline and the change back to default text format. If you navigate to before the newline and then hit return, your insertion point is _before_ the end `</p>` tag and the change back to default text format, hence you are then able to continue with the same format (and the newline then would be more like a `<br>` than a paragraph change).

Anyway .. yes, possibly that's not the same as a lot of word processors.. and there is just a special condition in the text layout code that essentially says "if we're read only (aka dynamic mode) and the last line of text is blank, just pretend there's one less line" so this is only affecting input mode text fields.

I did wonder about adding a new property for this... we have two options really (other than the "do nothing" approach!) which are:
1) add something using existing functionality using the SWF compatibility level, for example if you have `condenseWhite` set to true and have your SWF published with the latest version number, then we change the behaviour so that the extra newline is dropped in both cases rather than just for dynamic text fields
or
2) add a new property like you suggest above. Very simple to do, but an API change (and a challenge to think of a sensible yet concise name!). I'm actually wondering how best to set the behaviour because there may be instances where you want the newline to be there. So for example if you have `Start<p>para text</p>end` in HTML, then you'd get a line break at both the `<p>` and `</p>` - although in AIR it's only giving a newline at the closing `</p>` tag. Maybe then, we just need to make it a little more like HTML but with the understanding that a `<p>` at the start of a text field is just there to set the format and wouldn't cause a preceding line break, and a `</p>` at the end would be there just to close the format and wouldn't cause a terminating line break.

I'd be willing to accept any suggestions for naming such a property! So then we default it to the existing behaviour:
- starting `<p>` is never a line break and just sets the format; ending `</p>` is always a line break but a trailing blank line is hidden for dynamic text fields

and switch to the new behaviour if the property is explicitly set:
- `<p>` and `</p>` both cause a line break (as in HTML) except when they occur at the start or end of the `htmlText` respectively, at which point they're just signifying the format changes.

Maybe `paragraphBehaviourAsHtml` if that's not too much of a mouthful? But I'm very happy to accept other suggestions!!

thanks

**Reply by manishpatel0404** [reply]
If such a property is added, it would put to rest, an almost decade-old issue. More importantly, it would prevent recurrence of the issue being flagged in the future.

I struggled for a while trying to figure out a prospective name for the property and gave up. But your suggestion of `paragraphBehaviourAsHtml` seems appropriate, provided its use case is explained with an example in future reference docs.

Also, while we are on the subject of `TextField`, would you also please look into issue #1143 and #1979. I had to suspend work on issue #1143 because we do need to embed fonts and worked out a not-so-elegant temporary solution for issue #1979, while putting a proper solution (an ANE) on my ToDo list.

Thanks.

**Reply by ajwfrost** [reply]
Thanks - yes that sounds good then, we'll add this into the backlog.. documentation generation is pretty much there, so at some point soon we'll start publishing our own version of the AS3 docs (at least for AIR..)

I'll comment on those other issues within their threads.

thanks!

**Reply by manishpatel0404** [reply]
Great. Once this feature has been added it would be nice if the SO article is updated, as well. Great way to tell the world that AIR is Alive, Well and Kicking really hard...

Thanks.

**Comment by ylazy** [comment]
@ajwfrost Hi! I see the extra line break is still there. I'm using the latest Air SDK (50.1.1.2)

```as
var tf:TextField = new TextField();
tf.htmlParagraphBehavior = true;
tf.width = 400;
tf.height = 300;
tf.border = true;
tf.type = TextFieldType.INPUT;
tf.textColor = 0xFFFFFF;
tf.borderColor = 0xFFFFFF;
tf.multiline = true;
tf.htmlText = "<p><i>the quick blue dog</i></p>";
this.addChild(tf);
```
![image](https://user-images.githubusercontent.com/6010183/209603375-93d4e2f9-ac95-4f60-8cfc-36972dbbed45.png)

related issue: #2191. Please take a look at this.

Thanks!

**Reply by manishpatel0404** [reply]
Try setting `tf.htmlParagraphBehavior = false;` instead of `true`

**Reply by ylazy** [reply]
The same result:

![image](https://user-images.githubusercontent.com/6010183/209775972-0d9c163b-2808-45a7-8a0d-0d0aee1c06e6.png)

**Reply by ylazy** [reply]
[htmlParagraphBehavior docs:](https://airsdk.dev/reference/actionscript/3.0/flash/text/TextField.html#htmlParagraphBehavior)
> A Boolean value that determines whether `<p>` paragraph breaks should behave like HTML paragraphs. The default value is `false` which leaves the behaviour as it has been in AIR historically. Setting this to `true` will mean that any section of text marked as a paragraph, when embedded in a longer string, will be displayed with line breaks before and after it. However, line breaks will not be added if the start or end of a paragraph section coincides with the start or end of the text field.
> 
> The `htmlParagraphBehavior` property only affects text set with the `htmlText` property, not the text property. If you set text with the `text` property, `htmlParagraphBehavior` is ignored.
> 
> Set the `htmlParagraphBehavior` property before setting the `htmlText` property.

I believe `htmlParagraphBehavior` shouldn't behavior like that.
After many test, I think [this](https://github.com/airsdk/Adobe-Runtime-Support/issues/2191#issuecomment-1365592441) is the better behavior for `htmlParagraphBehavior`

**Reply by manishpatel0404** [reply]
Hmm. The `htmlParagraphBehavior` property was supposed to have been added to address the exact issue of the extra line break. From your example, it seems that the property is not doing what it is intended to do.

**Reply by ylazy** [reply]
As I understand the docs, the `htmlParagraphBehavior` is working like this:

![image](https://user-images.githubusercontent.com/6010183/209791666-d7714987-8069-4706-8213-a72a90d1232c.png)
![image](https://user-images.githubusercontent.com/6010183/209791683-89553aa6-301d-45d0-a4dc-060e27c4b748.png)

So that is more like a fix than a new feature.

**Reply by manishpatel0404** [reply]
> However, line breaks will not be added if the start or end of a paragraph section coincides with the start or end of the text field.

This is what does not seem to have been fixed in your original example.
